### PR TITLE
Feature archived

### PR DESCRIPTION
**Pull request by Snehdeep-ts**
# Description
Enhanced the archived section with hard delete functionality across all modules (users, roadmaps, courses, and lessons), improved UI components, loading states, and error handling. This change replaces the existing soft delete mechanism with a more efficient hard delete implementation while maintaining data integrity and includes significant UI/UX improvements for better visual consistency.

## Key Changes
1. Added Hard Delete Implementation
- Implemented permanent deletion endpoints for all modules
- Enhanced error handling with proper validations
- Added transaction support for data integrity
- Updated entity relationships and cascade behaviors
- Implemented proper cleanup of related entities

2. UI/UX Improvements
- Replaced full-page loader with skeleton loading states
- Added dedicated empty state component with contextual messaging
- Fixed button alignment and spacing issues
- Enhanced success/error notifications
- Improved confirmation modals
- Added pulse animation for better visual feedback
- Fixed modal vertical alignment and positioning
- Adjusted vertical spacing between search bar and content
- Improved search box component height and responsiveness
- Standardized text styling and padding in modal dialogs
- Enhanced cell layout and spacing consistency

3. Code Organization
- Moved static strings to localization file (en.ts)
- Created reusable components (EmptyState, EmptyStateIcons)
- Enhanced TypeScript typing
- Standardized component structure across archived sections
- Improved API service integration
- Implemented proper CSS organization using Tailwind classes

4. Entity Relationship Updates
```typescript
// Updated cascade behaviors
@ManyToOne(() => CourseEntity, (course) => course.lessons, {
  onDelete: 'CASCADE',
})
@ManyToOne(() => UserEntity, {
  onDelete: 'SET NULL',
})
@ManyToMany(() => RoadmapEntity, (roadmap) => roadmap.courses, {
  onDelete: 'CASCADE',
})
```

5. API Changes
- Updated DELETE endpoints for all modules:
```typescript
DELETE /users/:uuid      
DELETE /roadmap/:id      
DELETE /course/:id       
DELETE /lesson/:id       
```
- Enhanced error responses
- Added proper validation before deletion
- Updated API documentation

## Affected Components
- ArchivedCell (Updated):
  - Fixed cell layout and spacing
  - Improved button alignment
  - Enhanced text truncation
  - Standardized padding and margins
  - Optimized responsive behavior

- ConformationModal (Updated):
  - Fixed vertical alignment issues
  - Improved content spacing
  - Enhanced button layout
  - Better visual hierarchy
  - Standardized dialog positioning

- SearchBox (Updated):
  - Reduced excessive vertical padding
  - Improved search icon alignment
  - Better input field height
  - Enhanced visual consistency

- ArchivedUsers
- ArchivedRoadmaps
- ArchivedCourses
- ArchivedLessons
- EmptyState (new)
- EmptyStateIcons (new)
- Related Entities and Services
- API Controllers
- Database Schemas

## Migration Plan
1. Backup existing data
2. Update schema and constraints:
   - Remove deleted_at columns
   - Update foreign key constraints
   - Modify column nullability
3. Deploy backend changes:
   - Update entities
   - Add delete endpoints
   - Enhance error handling
4. Deploy frontend changes:
   - Update archived components
   - Add confirmation modals
   - Implement API integration
   - Apply UI/UX improvements

## Security Considerations
- Added validation before permanent deletion
- Implemented proper error handling
- Enhanced user feedback
- Added confirmation steps
- Maintained data integrity
- Added audit logging for delete operations

## Testing Checklist
- [ ] Verify all delete endpoints
- [ ] Test confirmation modals
- [ ] Validate success/error notifications
- [ ] Check cascade behaviors
- [ ] Test list refresh
- [ ] Verify loading states
- [ ] Test empty states
- [ ] Check data integrity
- [ ] Test transaction rollback
- [ ] Validate user permissions
- [ ] Verify modal vertical centering
- [ ] Check search box spacing and height
- [ ] Test responsive layout
- [ ] Validate text truncation in cells
- [ ] Verify UI consistency across all modules

## Documentation Updates
- Updated API documentation
- Added component documentation
- Updated entity relationship diagrams
- Added testing guidelines
- Updated migration guides
- Added UI/UX improvement documentation

### CODE REVIEW

File: apps/quick-learn-backend/src/entities/course.entity.ts [hunk 1]
```diff
@@ -20,7 +20,7 @@ export class CourseEntity extends BaseEntity {
   @Column({ type: 'varchar', length: 5000 })
   description: string;
 
-  @Column({ type: 'int', nullable: false })
+  @Column({ type: 'int', nullable: true })
```
Comment (shivraj-bhagat) [inline apps/quick-learn-backend/src/entities/course.entity.ts:23]: This shouldn't be nullable. Course category be deleted only when the user unassigned or changes the of each course it has.

File: apps/quick-learn-backend/src/routes/lesson/lesson.service.ts [hunk 2]
```diff
@@ -204,4 +208,33 @@ export class LessonService extends PaginationService<LessonEntity> {
       },
     );
   }
+
+  /**
+   * Permanently deletes a lesson
+   * @param id - The id of the lesson to delete
+   * @throws BadRequestException if the lesson doesn't exist
+   */
+  async deleteLesson(id: number): Promise<void> {
```
Comment (shivraj-bhagat) [inline apps/quick-learn-backend/src/routes/lesson/lesson.service.ts:217]: A transaction is not needed here. It is needed only when performing multiple operations, and two or more are related to each other.

File: apps/quick-learn-backend/src/routes/course/course.service.ts [hunk 3]
```diff
@@ -267,4 +267,15 @@ export class CourseService extends BasicCrudService<CourseEntity> {
       limit,
     };
   }
+
+  async deleteCourse(id: number): Promise<void> {
```
Comment (shivraj-bhagat) [inline apps/quick-learn-backend/src/routes/course/course.service.ts:271]: Also, handle the deletion of the lessons which are related to the course.

File: apps/quick-learn-backend/src/routes/roadmap/roadmap.service.ts [hunk 4]
```diff
@@ -278,4 +284,72 @@ export class RoadmapService extends PaginationService<RoadmapEntity> {
...
+    }
+
+    const roadmap = await this.getRoadmapById(id);
+
+    if (!roadmap) {
+      throw new BadRequestException(en.RoadmapNotFound);
+    }
+
+    // Start a transaction to ensure data consistency
+    const result = await this.roadmapRepository.manager.transaction(
```
Comment (shivraj-bhagat) [inline apps/quick-learn-backend/src/routes/roadmap/roadmap.service.ts:325]: Check for rollback and commit.

File: apps/quick-learn-frontend/tsconfig.json [hunk 5]
```diff
@@ -5,7 +5,7 @@
     "allowJs": true,
     "esModuleInterop": true,
     "allowSyntheticDefaultImports": true,
-    "strict": true,
+    "strict": false,
```
Comment (shivraj-bhagat) [inline apps/quick-learn-frontend/tsconfig.json:8]: Keep it default.

File: apps/quick-learn-frontend/src/shared/components/ArchivedCell.tsx [hunk 6]
```diff
@@ -20,64 +41,60 @@ const ArchivedCell: FC<Props> = ({
...
-          </button>
-          <button
-            className="text-gray-500 hover:text-red-600 hover:underline text-sm"
-            onClick={onClickDelete}
-          >
-            Delete
-          </button>
+        <div className="mt-2">
+          <div className="flex items-center text-sm text-gray-500">
+            <svg
```
Comment (shivraj-bhagat) [inline apps/quick-learn-frontend/src/shared/components/ArchivedCell.tsx:61]: Please move this to the UIElements.

File: apps/quick-learn-backend/src/routes/course/course.service.ts (hunk 3 above)
Comment (Snehdeep-ts) [reply apps/quick-learn-backend/src/routes/course/course.service.ts:271]: Looking at the lesson entity, I notice that we have onDelete: 'CASCADE' set on the course relationship, which means when a course is deleted, all its associated lessons will be automatically deleted by the database.

File: apps/quick-learn-backend/src/routes/lesson/lesson.service.ts [hunk 7]
```diff
@@ -8,12 +8,16 @@ import { en } from '@src/lang/en';
...
 import { PaginationDto } from '../users/dto';
 import { PaginatedResult } from '@src/common/interfaces';
+import { DataSource, Repository } from 'typeorm';
 
 @Injectable()
 export class LessonService extends PaginationService<LessonEntity> {
   constructor(
     @InjectRepository(LessonEntity) repo,
     private courseService: CourseService,
+    @InjectRepository(LessonEntity)
```
Comment (shivraj-bhagat) [inline apps/quick-learn-backend/src/routes/lesson/lesson.service.ts:18]: The pagination service exposes a repository which is the same as here injecting it

File: apps/quick-learn-frontend/src/app/(Dashboard)/dashboard/archived/courses/ArchivedCourses.tsx [hunk 8]
```diff
@@ -10,23 +10,42 @@ import React, {
...
-import { FullPageLoader } from '@src/shared/components/UIElements';
 import { debounce } from '@src/utils/helpers';
 import InfiniteScroll from 'react-infinite-scroll-component';
 import ConformationModal from '@src/shared/modals/conformationModal';
 import { TCourse } from '@src/shared/types/contentRepository';
 import { en } from '@src/constants/lang/en';
 import { toast } from 'react-toastify';
+import EmptyState from '@src/shared/components/EmptyStatePlaceholder';
+
+const LoadingSkeleton = () => (
```
Comment (shivraj-bhagat) [inline apps/quick-learn-frontend/src/app/(Dashboard)/dashboard/archived/courses/ArchivedCourses.tsx:25]: Make a different component

File: apps/quick-learn-frontend/src/shared/components/SearchBox.tsx [hunk 9]
```diff
@@ -25,19 +24,18 @@ const SearchBox: FC<Props> = ({ value, handleChange }) => {
           >
```
Comment (shivraj-bhagat) [inline apps/quick-learn-frontend/src/shared/components/SearchBox.tsx:24]: make this in a separate file as we already have the UI elements.

File: apps/quick-learn-frontend/src/shared/modals/conformationModal.tsx [hunk 10]
```diff
@@ -28,7 +28,12 @@ export default function ConformationModal({
   };
 
   return (
-    <Modal show={open} size="md" popup>
+    <Modal
+      show={open}
+      size="md"
+      popup
```
Comment (shivraj-bhagat) [inline apps/quick-learn-frontend/src/shared/modals/conformationModal.tsx:34]: clean this up

**Review by shivraj-bhagat — APPROVED** [review]
lgtm

**Pull request merged**